packages/apps/job-launcher/server/src/modules/job/routing-protocol.service.ts

```diff
@@ -1,34 +1,208 @@
-    this.priorityOrder = this.generatePriorityOrder();
+    this.reputationOracles = this.web3ConfigService.reputationOracles
+      .split(',')
+      .map((address) => address.trim());
+
+    this.chainPriorityOrder = this.shuffleArray(this.chains);
+    this.reputationOraclePriorityOrder = this.shuffleArray(
+      this.reputationOracles.map((_, i) => i),
+    );
+
+    this.oracleOrder = this.chains.reduce((acc: OracleOrder, chainId) => {
+      acc[chainId] = this.reputationOracles.reduce(
+        (oracleAcc, reputationOracle) => {
+          oracleAcc[reputationOracle] = {
+            [Role.ExchangeOracle]: [],
+            [Role.RecordingOracle]: [],
+          };
+          return oracleAcc;
+        },
+        {} as { [reputationOracle: string]: { [role: string]: string[] } },
+      );
+      return acc;
+    }, {} as OracleOrder);
   }
 
-  private generatePriorityOrder() {
-    const priorityOrder = Array.from(Array(this.chains.length).keys());
-    for (let i = priorityOrder.length - 1; i > 0; i--) {
-      const j = Math.floor(Math.random() * (i + 1));
-      [priorityOrder[i], priorityOrder[j]] = [
-        priorityOrder[j],
-        priorityOrder[i],
-      ];
-    }
-    return priorityOrder;
+  private shuffleArray<T>(array: T[]): T[] {
+    return array.sort(() => Math.random() - 0.5);
   }
 
   public selectNetwork(): ChainId {
-    const chainId = this.chains[this.priorityOrder[this.currentIndex]];
-    this.currentIndex = (this.currentIndex + 1) % this.chains.length;
+    const chainId =
+      this.chains[this.chainPriorityOrder[this.chainCurrentIndex]];
+    this.chainCurrentIndex = (this.chainCurrentIndex + 1) % this.chains.length;
     return chainId;
   }
+
+  public selectReputationOracle(): string {
+    const reputationOracle =
+      this.reputationOracles[
+        this.reputationOraclePriorityOrder[this.reputationOracleIndex]
+      ];
+
+    this.reputationOracleIndex =
+      (this.reputationOracleIndex + 1) % this.reputationOracles.length;
+    return reputationOracle;
+  }
+
+  public async selectOracleFromAvailable(
+    availableOracles: any[],
+    oracleType: string,
+    chainId: ChainId,
+    reputationOracle: string,
+  ): Promise<string | null> {
+    const oraclesOfType = availableOracles
+      .filter((oracle) => oracle.role === oracleType)
+      .map((oracle) => oracle.address);
+
+    if (!oraclesOfType.length) {
+      return null;
+    }
+
+    const latestOraclesHash = hashString(JSON.stringify(oraclesOfType));
+    // Check if we need to shuffle and store the new order
+    if (this.oraclesHash !== latestOraclesHash) {
+      this.oraclesHash = latestOraclesHash;
+      const shuffledOracles = this.shuffleArray(oraclesOfType);
+      this.oracleOrder[chainId][reputationOracle][oracleType] = shuffledOracles;
+    }
+
+    const orderedOracles =
+      this.oracleOrder[chainId][reputationOracle][oracleType];
+
+    const currentIndex = this.reputationOracleIndex % orderedOracles.length;
```

Comment:
You are using the reputation oracle index to select other oracles type.
To be totally fair we would need an index per reputation network & oracle type & job type, but it would make things too complex.
So let's change the logic and select the oracle randomly from the list of available oracles based on the reputation network & oracle type & job type

Reply:
Yes, I wanted to implement a more complex choice in which the same oracle cannot be selected twice in a round, for example, when we use randomization. That is, now we have indices, as in network selection.